src/Orchard.Web/Modules/Orchard.OpenId/InMemoryCache.cs

```diff
@@ -1,13 +1,12 @@
 ﻿using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
-using System.Threading;
 using Microsoft.IdentityModel.Clients.ActiveDirectory;
 
 namespace Orchard.OpenId {
     public class InMemoryCache : TokenCache {
-        private static Dictionary<string, byte[]> _inMemoryTokenCache;
-        private static ReaderWriterLockSlim _sessionLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+        private static ConcurrentDictionary<string, byte[]> _inMemoryTokenCache;
```

Comment:
The keys of the dictionary are unique user ids (the id of the user in the open id provider system, not Orchard's id), so even across multiple tenants the dictionary will be safe.